ChopChop/Models/Ingredients/Ingredient.swift

```diff
@@ -0,0 +1,271 @@
+        }
+
+        let currentDate = Date()
+        var subtractedQuantity = quantity
+        var usedBatches: [IngredientBatch] = []
+
+        for batch in batchesByExpiryDate {
+            guard let expiryDate = batch.expiryDate else {
+                do {
+                    try subtractedQuantity -= batch.quantity
+                    usedBatches.append(batch)
+                    continue
+                } catch QuantityError.negativeQuantity {
+                    try batch.subtract(subtractedQuantity)
+                    subtractedQuantity.value = 0
+                    break
+                }
+            }
+
+            guard expiryDate > currentDate else {
+                continue
+            }
+
+            do {
+                try subtractedQuantity -= batch.quantity
+                usedBatches.append(batch)
+            } catch QuantityError.negativeQuantity {
+                try batch.subtract(subtractedQuantity)
+                subtractedQuantity.value = 0
+                break
+            }
+        }
+
+        guard subtractedQuantity.value == 0 else {
+            throw IngredientError.insufficientQuantity
+        }
+
+        removeBatches(usedBatches)
+    }
+
+    /**
+     Combines this ingredient with another of the same name and quantity type.
+     Batches with the same expiry date are merged.
+     - Throws:
+        - `QuantityError.differentQuantityTypes`:
+            if the given quantity does not have the same type as the ingredient.
+        - `QuantityError.negativeQuantity`:
+            f any addition result is negative.
+        - `IngredientError.differentIngredients`:
+            if the given ingredient does not have the same name as this ingredient.
+     */
+    func combine(with ingredient: Ingredient) throws {
+        guard self.name == ingredient.name else {
+            throw IngredientError.differentIngredients
+        }
+
+        guard self.quantityType == ingredient.quantityType else {
+            throw QuantityError.differentTypes
+        }
+
+        for batch in ingredient.batches {
+            let quantity = batch.quantity
+            let expiryDate = batch.expiryDate
+            try add(quantity: quantity, expiryDate: expiryDate)
+        }
+    }
+}
+
+// MARK: - Batch operations
+extension Ingredient {
+    /**
+     Returns the batch with the given expiry date, or `nil` if it does not exist.
+     */
+    func getBatch(expiryDate: Date?) -> IngredientBatch? {
+        batches.first { batch in
+            batch.expiryDate == expiryDate
+        }
+    }
+
+    /**
+     Removes the batch identified by the given expiry date.
+     If such a batch does not exist, do nothing.
+     */
+    func removeBatch(expiryDate: Date?) {
+        batches.removeAll { batch in
+            batch.expiryDate == expiryDate
+        }
+    }
+
+    /**
+     Removes all batches with expiry dates earlier than the current date.
+     */
+    func removeExpiredBatches() {
```

Comment:
interesting, are we gonna call this method periodically from somewhere?

Reply:
This seemed like a useful function to have (maybe there will be a button that users can press to rid an ingredient of all its expired batches)